pkg/io/buffered.go

```diff
@@ -0,0 +1,283 @@
+	var buf *readerBuffer
+	for i := range r.buffers {
+		if r.canRead(&r.buffers[i], offset, int64(len(b))) {
+			buf = &r.buffers[i]
+			break
+		}
+
+		if lru == nil || r.buffers[i].count < lru.count {
+			lru = &r.buffers[i]
+		}
+	}
+
+	if buf == nil {
+		// No buffer satisfied read, overwrite least-recently-used
+		buf = lru
+		r.prep(buf, offset, int64(len(b)))
+
+		// Here we exchange the top-level lock for
+		// the buffer's individual write lock
+		buf.mtx.Lock()
+		defer buf.mtx.Unlock()
+		buf.count = r.count
+		r.mtx.Unlock()
+
+		if _, err := r.populate(buf); err != nil {
+			return 0, err
+		}
+
+		r.read(buf, b, offset)
+		return len(b), nil
+	}
+
+	// Here we exchange the top-level lock for
+	// the buffer's individual read lock
+	buf.mtx.RLock()
+	defer buf.mtx.RUnlock()
+	buf.count = r.count
+	r.mtx.Unlock()
+
+	r.read(buf, b, offset)
+	return len(b), nil
+}
+
+type BufferedWriter struct {
+	w   io.Writer
+	buf []byte
+}
+
+func NewBufferedWriter(w io.Writer) BufferedWriteFlusher {
+	return &BufferedWriter{w, nil}
+}
+
+var _ BufferedWriteFlusher = (*BufferedWriter)(nil)
+
+func (b *BufferedWriter) Write(p []byte) (n int, err error) {
+	b.buf = append(b.buf, p...)
+	return len(p), nil
+}
+
+func (b *BufferedWriter) Len() int {
+	return len(b.buf)
+}
+
+func (b *BufferedWriter) Flush() error {
+	n, err := b.w.Write(b.buf)
+	b.buf = b.buf[:len(b.buf)-n]
+	return err
+}
+
+func (b *BufferedWriter) Close() error {
+	if len(b.buf) > 0 {
+		err := b.Flush()
+		b.buf = nil
+		return err
+	}
+	return nil
+}
+
+type BufferedWriteFlusher interface {
+	io.WriteCloser
+	Len() int
+	Flush() error
+}
+
+// BufferedWriterWithQueue is an attempt at writing an async queue of outgoing data to the underlying writer.
+// As a tradeoff for removing flushes from the hot path, this writer does not provide guarantees about
+// bubbling up errors and takes a best effort approach to signal a failure.
+type BufferedWriterWithQueue struct {
+	w   io.Writer
+	buf []byte
+
+	flushCh chan []byte
+	doneCh  chan struct{}
+	err     atomic.Error
+}
+
+var _ BufferedWriteFlusher = (*BufferedWriterWithQueue)(nil)
+
+func NewBufferedWriterWithQueue(w io.Writer) BufferedWriteFlusher {
+	b := &BufferedWriterWithQueue{
+		w:       w,
+		buf:     nil,
+		flushCh: make(chan []byte, 10), // todo: guess better?
+		doneCh:  make(chan struct{}),
+	}
+
+	go b.flushLoop()
+
+	return b
+}
+
+func (b *BufferedWriterWithQueue) Write(p []byte) (n int, err error) {
+	b.buf = append(b.buf, p...)
+	return len(p), nil
+}
+
+func (b *BufferedWriterWithQueue) Len() int {
+	return len(b.buf)
+}
+
+func (b *BufferedWriterWithQueue) Flush() error {
+	if err := b.err.Load(); err != nil {
+		return errors.Wrap(err, "error in async write using buffered writer")
+	}
+
+	bufCopy := make([]byte, 0, len(b.buf))
+	bufCopy = append(bufCopy, b.buf...)
+
+	// reset/resize buffer
+	b.buf = b.buf[:0]
+
+	// will only block if the entire buffered channel is full
+	b.flushCh <- bufCopy
```

Comment:
i don't know enough about what is relying on this code to be sure but is blocking preferred here? should we do a channel select with a default and return error if the queue is full? Just want to make sure that blocking here won't lock compaction or something.

Reply:
The goal is to have at least some concurrency and improve compaction performance.  I'd say limited buffer and blocking is the preferred behavior.  If compaction input exceeds the flush speed to the backend then it will block until more room is available.  We should be able to rely on the flush loop to timeout and error (it won't try indefinitely).